
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../call_spots/">
      
      
        <link rel="next" href="../troubleshoot/">
      
      
      <link rel="icon" href="../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>OMP - Coppafisher Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/extra.css">
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#orthogonal-matching-pursuit-omp" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Coppafisher Documentation" class="md-header__button md-logo" aria-label="Coppafisher Documentation" data-md-component="logo">
      
  <img src="../images/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Coppafisher Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              OMP
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/paulshuker/coppafisher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Coppafisher
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Getting started

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../basic_usage/" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../diagnostics/" class="md-tabs__link">
        
  
  
    
  
  Diagnostics

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../export/" class="md-tabs__link">
        
  
  
    
  
  Export

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../overview/" class="md-tabs__link">
          
  
  
  Method

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../troubleshoot/" class="md-tabs__link">
        
  
  
    
  
  Troubleshoot

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../glossary/" class="md-tabs__link">
        
  
  
    
  
  Glossary

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../migration/" class="md-tabs__link">
        
  
  
    
  
  Migration

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../contributing/" class="md-tabs__link">
        
  
  
    
  
  Contributing

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Coppafisher Documentation" class="md-nav__button md-logo" aria-label="Coppafisher Documentation" data-md-component="logo">
      
  <img src="../images/logo.svg" alt="logo">

    </a>
    Coppafisher Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/paulshuker/coppafisher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Coppafisher
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic_usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic Usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced_usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Advanced Usage
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../diagnostics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Diagnostics
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../export/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Export
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Method
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Method
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../find_spots/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Find Spots
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../stitch/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Stitch
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../register/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Registration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../call_spots/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Call Spots
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    OMP
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    OMP
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#definitions" class="md-nav__link">
    <span class="md-ellipsis">
      Definitions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#0-pre-processing" class="md-nav__link">
    <span class="md-ellipsis">
      0: Pre-processing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-minimum-intensity-threshold" class="md-nav__link">
    <span class="md-ellipsis">
      1: Minimum Intensity Threshold
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-next-gene-assignment" class="md-nav__link">
    <span class="md-ellipsis">
      2: Next Gene Assignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-gene-weights" class="md-nav__link">
    <span class="md-ellipsis">
      3: Gene Weights
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-pixel-scores" class="md-nav__link">
    <span class="md-ellipsis">
      4: Pixel Scores
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-spot-scoring-and-spot-detection" class="md-nav__link">
    <span class="md-ellipsis">
      5: Spot Scoring and Spot Detection
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#diagnostics" class="md-nav__link">
    <span class="md-ellipsis">
      Diagnostics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Diagnostics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#intensity" class="md-nav__link">
    <span class="md-ellipsis">
      Intensity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#viewer" class="md-nav__link">
    <span class="md-ellipsis">
      Viewer
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pdf" class="md-nav__link">
    <span class="md-ellipsis">
      PDF
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshoot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshoot
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../glossary/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Glossary
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../migration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Migration
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#definitions" class="md-nav__link">
    <span class="md-ellipsis">
      Definitions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#0-pre-processing" class="md-nav__link">
    <span class="md-ellipsis">
      0: Pre-processing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-minimum-intensity-threshold" class="md-nav__link">
    <span class="md-ellipsis">
      1: Minimum Intensity Threshold
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-next-gene-assignment" class="md-nav__link">
    <span class="md-ellipsis">
      2: Next Gene Assignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-gene-weights" class="md-nav__link">
    <span class="md-ellipsis">
      3: Gene Weights
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-pixel-scores" class="md-nav__link">
    <span class="md-ellipsis">
      4: Pixel Scores
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-spot-scoring-and-spot-detection" class="md-nav__link">
    <span class="md-ellipsis">
      5: Spot Scoring and Spot Detection
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#diagnostics" class="md-nav__link">
    <span class="md-ellipsis">
      Diagnostics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Diagnostics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#intensity" class="md-nav__link">
    <span class="md-ellipsis">
      Intensity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#viewer" class="md-nav__link">
    <span class="md-ellipsis">
      Viewer
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pdf" class="md-nav__link">
    <span class="md-ellipsis">
      PDF
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="orthogonal-matching-pursuit-omp">Orthogonal Matching Pursuit (OMP)</h1>
<p>OMP is coppafisher's current best gene assignment algorithm. OMP runs independently, except requiring
<a href="../overview/#register">register</a> for image-alignment and <a href="../overview/#call-spots">call spots</a> for dataset-accurate
representation of each gene's unique barcode: its bled code, <span class="arithmatex">\(\mathbf{B}\)</span>.</p>
<p>A pixel score image is produced for every pixel and every gene by iterating through steps 2-4. Then, the final gene
reads are found in <a href="#5-spot-scoring-and-spot-detection">step 5</a>.</p>
<h2 id="definitions">Definitions</h2>
<ul>
<li><span class="arithmatex">\(r\)</span> and <span class="arithmatex">\(c\)</span> represents sequencing rounds and channels respectively.</li>
<li><span class="arithmatex">\(B_{grc}\)</span> represents gene g's bled code in round <span class="arithmatex">\(r\)</span>, channel <span class="arithmatex">\(c\)</span> saved at <code>nb.call_spots.bled_codes</code> from call spots.</li>
<li><span class="arithmatex">\(S_{prc}\)</span> is pixel <span class="arithmatex">\(p\)</span>'s colour in round <span class="arithmatex">\(r\)</span>, channel <span class="arithmatex">\(c\)</span>, after pre-processing is applied.</li>
<li><span class="arithmatex">\(c_{pgi}\)</span> is the OMP pixel score given to gene <span class="arithmatex">\(g\)</span> at pixel <span class="arithmatex">\(p\)</span> on the <span class="arithmatex">\(i\)</span>'th OMP iteration.</li>
<li><span class="arithmatex">\(w_{pgi}\)</span> is the OMP gene weight given to gene <span class="arithmatex">\(g\)</span> for image pixel <span class="arithmatex">\(p\)</span> on the <span class="arithmatex">\(i\)</span>'th iteration. This is computed by
least squares in <a href="#3-gene-weights">step 3</a>. <span class="arithmatex">\(i\)</span> takes values <span class="arithmatex">\(1, 2, 3, ...\)</span></li>
<li><span class="arithmatex">\(||A||_{...}\)</span> represents an L2 norm of <span class="arithmatex">\(A\)</span> (or Frobenius norm for a matrix) over all indices replaced by a dot (<span class="arithmatex">\(.\)</span>).</li>
</ul>
<h2 id="0-pre-processing">0: Pre-processing</h2>
<p>All pixel colours are gathered using the results from register. Any out of bounds round/channel colour intensities are
set to zero. The pixel colours, <span class="arithmatex">\(\mathbf{S}\)</span>, are multiplied by <code>nb.call_spots.colour_norm_factor</code> for the each tile.</p>
<h2 id="1-minimum-intensity-threshold">1: Minimum Intensity Threshold</h2>
<p>Before running on pixels, many pixels are discarded because they are background and not spots. To do this, we take the
middle z plane colours for each tile, <span class="arithmatex">\(D_{txyrc}\)</span>, and compute their intensities as</p>
<div class="arithmatex">\[
I_{txy} = \min_r(\max_c(|D_{txyrc}|))
\]</div>
<p>The intensity thresholds for each tile are then</p>
<div class="arithmatex">\[
\text{minimum\_intensity}_t = a\times\text{nth percentile}_{xy}(I_{txy})
\]</div>
<p><span class="arithmatex">\(n\)</span> is the <code>minimum_intensity_percentile</code> (typically 5) and <span class="arithmatex">\(a\)</span> is the <code>minimum_intensity_multiplier</code> (typically 4).</p>
<h2 id="2-next-gene-assignment">2: Next Gene Assignment</h2>
<p>A pixel can have more than one gene assigned to it. The most genes allowed on each pixel is <code>max_genes</code>
(typically 5). Let's say we are on iteration <span class="arithmatex">\(i\)</span> (<span class="arithmatex">\(i = 1, 2, 3, ...\)</span>) for pixel <span class="arithmatex">\(p\)</span>. The pixel will already have
<span class="arithmatex">\(i - 1\)</span> genes assigned to it and their weights have been computed <span class="arithmatex">\((w_{pg(i - 1)})\)</span>. We compute the latest residual
pixel colour <span class="arithmatex">\(R_{prci}\)</span> as</p>
<div class="arithmatex">\[
R_{prci} = S_{prc} - \sum_g(w_{pg(i - 1)}B_{grc})
\]</div>
<p>For the first iteration, <span class="arithmatex">\(R_{prc(i=1)} = S_{prc}\)</span>. Using this residual, a "semi dot product score" is
computed for every gene and background gene <span class="arithmatex">\(g\)</span> similar to
<a href="../call_spots/#6-and-7-application-of-scales-computation-of-final-scores-and-bleed-matrix">call spots</a></p>
<div class="arithmatex">\[
\text{(gene scores)}_{pgi} = \frac{1}{N_r}\Bigg|\sum_r\Bigg(\frac{1}{||\mathbf{\hat{R}}||_{pr.(i-1)}||\mathbf{B}||_{gr.}}\sum_{c}(\hat{R}_{prc(i - 1)}B_{grc})\Bigg)\Bigg|
\]</div>
<p>where</p>
<div class="arithmatex">\[
\hat{R}_{prci} = \epsilon_{prci}^2R_{prci}\text{,}\space\space\space
\epsilon_{prci}^2 = N_r N_c\frac{\sigma_{pirc}^{-2}}{\sum_{rc}\sigma_{pirc}^{-2}}
\]</div>
<p>and</p>
<div class="arithmatex">\[
\sigma_{pirc}^2 = \beta^2 + \alpha\sum_{g\text{ assigned}} w_{pg(i-1)}^2 B_{grc}^2\text{,}\space\space\space
N_r = \sum_r 1\text{,}\space\space\space
N_c = \sum_c 1
\]</div>
<p><span class="arithmatex">\(\alpha\)</span> is given by <code>alpha</code> (typically 120) and boosts the uncertainty on round-channel pairs already strongly
weighted. <span class="arithmatex">\(\beta\)</span> is given by <code>beta</code> (typically 1) and gives every round-channel pair a constant uncertainty.</p>
<details class="question">
<summary>Why do we need an uncertainty weighting (<span class="arithmatex">\(\mathbf{\epsilon}^2\)</span>) for each round-channel pair?</summary>
<p>On real datasets, subtracting the assigned, weighted bled code is not perfect for every round (shown below).
Therefore, <span class="arithmatex">\(\mathbf{\epsilon}\)</span> is a way of estimating the uncertainty associated with the imperfect gene weights. It
places a bias towards genes that are bright in unique round-channel pairs when <span class="arithmatex">\(\alpha&gt;0\)</span>.</p>
<p>By default, <span class="arithmatex">\(\alpha &gt;&gt; \beta\)</span> so it is unlikely to assign two genes bright in the same round-channel pairs.</p>
<p>If you fully trust the weightings to be accurate, set <code>alpha</code> to zero in the config.</p>
<p><figure markdown="span">
  <img alt="Image title" height="300" src="../images/algorithm/omp/poor_gene_weight_example.png" />
  <figcaption>An example of OMP failing to find a scalar to correctly weight every bright round-channel pair for a
  gene. It is failing because the residual colour is sometimes very positive and sometimes very negative. Only the
  fourth round was almost perfectly subtracted.</figcaption>
</figure></p>
</details>
<p>Gene <span class="arithmatex">\(\tilde{g}\)</span> is successfully assigned to pixel <span class="arithmatex">\(p\)</span> when all conditions are met:</p>
<ul>
<li><span class="arithmatex">\((\text{gene scores})_{p\tilde{g}i}\)</span> is the largest scoring gene.</li>
<li><span class="arithmatex">\((\text{gene scores})_{p\tilde{g}i} &gt;\)</span> <code>dot_product_threshold</code> (typically 0.72).</li>
<li><span class="arithmatex">\(\tilde g\)</span> is not already assigned to the pixel.</li>
<li><span class="arithmatex">\(\tilde g\)</span> is not a background gene.</li>
<li>The residual intensity <span class="arithmatex">\(\min_r(\max_c(|\hat{R}_{prci}|)) &gt; \text{minimum\_intensity}_t\)</span>. See <a href="../diagnostics/#intensity-images">diagnostic</a>.</li>
<li>Iteration <span class="arithmatex">\(i \leq\)</span> <code>max_genes</code>.</li>
</ul>
<p>The reasons for each of these conditions is to:</p>
<ul>
<li>pick the best gene</li>
<li>remove unconfident gene reads</li>
<li>not double assign genes</li>
<li>avoid over-fitting on high-background pixel colour</li>
<li>remove dim colours (background noise)</li>
<li>avoid assigning too many genes</li>
</ul>
<p>respectively. If a pixel fails to meet one or more of these conditions, then no more genes are assigned to it and the
pixel scores will be final.</p>
<p>If pixel <span class="arithmatex">\(p\)</span> meets all conditions, then gene <span class="arithmatex">\(\tilde g\)</span> is taken as the next gene assigned.</p>
<p>If all remaining pixels fail the conditions, then the iterations stop and the current pixel scores <span class="arithmatex">\(\mathbf{c}\)</span> are kept
as final for <a href="#5-spot-scoring-and-spot-detection">step 5</a>.</p>
<h2 id="3-gene-weights">3: Gene Weights</h2>
<p>On each iteration, the gene weights are re-computed for all genes assigned to pixel <span class="arithmatex">\(p\)</span> to best represent the pixel's
colour. All unassigned genes have a zero weight, so <span class="arithmatex">\(g\)</span> here represents only the assigned genes (<span class="arithmatex">\(i\)</span> assigned genes)
for pixels that passed <a href="#2-next-gene-assignment">step 2</a>. The weights, <span class="arithmatex">\(w_{pgi}\)</span>, are computed through the
method of least squares by minimising the scalar residual</p>
<div class="arithmatex">\[
\sum_{rc}(S_{prc} - \sum_{g\text{ assigned}}(B_{grc}w_{pgi}))^2
\]</div>
<p>In other words, using matrix multiplication, the weight vector of length genes assigned is</p>
<div class="arithmatex">\[
\mathbf{w} = \bar{\mathbf{B}}^{-1} \bar{\mathbf{S}}
\]</div>
<p>where <span class="arithmatex">\(\bar{(...)}\)</span> represents flattening the round and channel dimensions into a single dimension, so
<span class="arithmatex">\(\bar{\mathbf{B}}\)</span> is of shape <span class="arithmatex">\(\text{genes assigned}\)</span> by <span class="arithmatex">\(\text{rounds}*\text{channels}\)</span> and <span class="arithmatex">\(\bar{\mathbf{S}}\)</span> is of
shape <span class="arithmatex">\(\text{rounds} * \text{channels}\)</span>. <span class="arithmatex">\((...)^{-1}\)</span> is the Moore-Penrose matrix inverse (a pseudo-inverse).</p>
<h2 id="4-pixel-scores">4: Pixel Scores</h2>
<p>After updating the gene weights, every assigned gene pixel score is (re)computed for pixels that passed gene assignment.
The pixel score for assigned gene <span class="arithmatex">\(g\)</span> in pixel <span class="arithmatex">\(p\)</span> is given by</p>
<div class="arithmatex">\[
c_{pgi} = \frac{1}{N_r ||\mathbf{\tilde{R}}||_{pgr.i}}\Bigg | \sum_{rc}\tilde{R}_{pgrci} \hat{B}_{grc} \Bigg |
\]</div>
<p>where</p>
<div class="arithmatex">\[
\tilde{R}_{pgrci} = \epsilon_{prci}^2\Bigg( S_{prc} - \sum_{g'\text{; }g' \neq g}B_{g'rc}w_{pg'i}\Bigg)
\]</div>
<p>and</p>
<div class="arithmatex">\[
\epsilon_{pgrci}^2 = N_r N_c \frac{\sigma_{pgirc}^{-2}}{\sum_{rc} \sigma_{pgirc}^{-2}} \text{,}\space\space\space
\sigma_{pgirc}^2 = \beta^2 + \alpha \sum_{g'\text{; }g' \neq g}w_{pg'i}^2 B_{g'rc}^2
\]</div>
<p>A pixel score is made negative if the gene's weight is negative.</p>
<p><a href="#2-next-gene-assignment">Step 2</a> is now repeated on the remaining pixels.</p>
<details class="question">
<summary>Why not use the scores from step 2 as the pixel scores?</summary>
<p>If you recall, from <a href="#2-next-gene-assignment">step 2</a>, the assigned gene is given a preliminary score similar to
step 4's pixel score. This score is not used as the final OMP pixel scores (but, we did try). This is because the
pleminary score has lowered the scores because they overlap with other genes. In other words, the scores are lowered
by brightness in other rounds-channel pairs.</p>
<p>The step 4 scoring method gets around this. Assuming that all gene assignments are perfect, by subtracting those
assignments off except gene g, then gene g is given a fairer chance of scoring highly, hopefully without the
brightness of other genes.</p>
</details>
<h2 id="5-spot-scoring-and-spot-detection">5: Spot Scoring and Spot Detection</h2>
<p>The gene pixel score images are converted to gene score images by convolving with the mean spot given as a numpy .npy
file at file path <code>omp_mean_spot</code> in the <code>file_names</code> config section. If <code>omp_mean_spot</code> is not given, the default mean
spot is used (shown below). The mean spot is divided by its sum then used. This gives a score for every pixel for every
gene. Spot-shaped and high pixel score regions result in higher score maxima. Scores can be <span class="arithmatex">\(\geq 0\)</span>. But, in practice,
scores are rarely greater than <span class="arithmatex">\(1\)</span>.</p>
<figure>
  <img alt="Image title" src="../images/algorithm/omp/omp_mean_spot_example.png" width="776" />
<br />
<figcaption>The default mean spot. The middle image is the central z plane.</figcaption>
</figure>
<p>Using the score images, each gene's spots are detected using the <a href="../find_spots/">find spots</a> algorithm to find score
local maxima using config parameters <code>radius_xy</code> (typically <code>3</code>) and <code>radius_z</code> (typically <code>2</code>) respectively with a
score threshold set by <code>score_threshold</code> (typically <code>0.1</code>). These are the final OMP gene reads shown in the
<a href="../diagnostics/#viewer">Viewer</a>.</p>
<details class="question">
<summary>Why not score each spot using a single pixel score value?</summary>
<p>Pixel scores can be inflated by single overly-bright round/channel anomalies since they are computed using a
non-robust least squares calculation. This could be from real autofluorescence or from mistakes in registration.
For this reason, a spot's score is better represented by using pixel score data from its neighbourhood. The mean
spot is an estimation of how much care to put in the local, spatial neighbourhood.</p>
<p>If you still wanted to score each spot by a single pixel score value, create your own 1x1x1 mean spot with value
<span class="arithmatex">\(&gt;0\)</span> and run OMP again.</p>
</details>
<h2 id="diagnostics">Diagnostics</h2>
<h3 id="intensity">Intensity</h3>
<p>It is recommended to view the pixel intensities on a histogram first by typing</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">coppafisher</span><span class="w"> </span><span class="kn">import</span> <span class="n">Notebook</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">coppafisher.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">intensity</span>

<span class="n">nb</span> <span class="o">=</span> <span class="n">Notebook</span><span class="p">(</span><span class="s2">&quot;/path/to/notebook&quot;</span><span class="p">)</span>
<span class="n">intensity</span><span class="o">.</span><span class="n">view_intensity_histogram</span><span class="p">(</span><span class="n">nb</span><span class="p">)</span>
</code></pre></div>
<p>On this plot, you can see vertical lines to indicate important intensity values.</p>
<p>Use the <a href="../diagnostics/#intensity-images">diagnostic</a> to see intensity images. Once OMP is complete, you can view the
minimum intensity computed for each tile by doing</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">coppafisher</span><span class="w"> </span><span class="kn">import</span> <span class="n">Notebook</span>

<span class="n">nb</span> <span class="o">=</span> <span class="n">Notebook</span><span class="p">(</span><span class="s2">&quot;/path/to/notebook&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nb</span><span class="o">.</span><span class="n">omp</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;tile_</span><span class="si">{</span><span class="n">tile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;minimum_intensity&quot;</span><span class="p">])</span>
</code></pre></div>
<p>where <code>tile</code> is the tile index.</p>
<h3 id="viewer">Viewer</h3>
<p>Use the <a href="../diagnostics/#viewer">Viewer</a> to check the final gene reads made by OMP.</p>
<h3 id="pdf">PDF</h3>
<p>Check the <code>_omp.pdf</code> file created at runtime in the output directory for details on the OMP mean spot, the spot score
distributions, gene counts, and gene locations.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>